src/common/Utilities/StringConvert.h

```diff
@@ -171,6 +170,91 @@ namespace Trinity::Impl::StringConvertImpl
             return (val ? "1" : "0");
         }
     };
+
+#if TRINITY_COMPILER == TRINITY_COMPILER_MICROSOFT
+    template <typename T>
+    struct For<T, std::enable_if_t<std::is_floating_point_v<T>>>
+    {
+        static Optional<T> FromString(std::string_view str, std::chars_format fmt = std::chars_format())
+        {
+            if (str.empty())
+                return std::nullopt;
+
+            if (fmt == std::chars_format())
+            {
+                if (StringEqualI(str.substr(0, 2), "0x"))
+                {
+                    fmt = std::chars_format::hex;
+                    str.remove_prefix(2);
+                }
+                else
+                    fmt = std::chars_format::general;
+
+                if (str.empty())
+                    return std::nullopt;
+            }
+
+            char const* const start = str.data();
+            char const* const end = (start + str.length());
+
+            T val;
+            std::from_chars_result const res = std::from_chars(start, end, val, fmt);
+            if ((res.ptr == end) && (res.ec == std::errc()))
+                return val;
+            else
+                return std::nullopt;
+        }
+
+        // this allows generic converters for all numeric types (easier templating!)
+        static Optional<T> FromString(std::string_view str, int base)
+        {
+            if (base == 16)
+                return FromString(str, std::chars_format::hex);
+            else if (base == 10)
+                return FromString(str, std::chars_format::general);
+            else
+                return FromString(str, std::chars_format());
+        }
+
+        static std::string ToString(T val)
+        {
+            return std::to_string(val);
+        }
+    };
+#else
+    // @todo replace this once libc++ supports double args to from_chars
+    template <typename T>
+    struct For<T, std::enable_if_t<std::is_floating_point_v<T>>>
+    {
+        static Optional<T> FromString(std::string_view str, int base = 0)
+        {
+            try {
+                if (str.empty())
+                    return std::nullopt;
+
+                if ((base == 10) && StringEqualI(str.substr(0, 2), "0x"))
+                    return std::nullopt;
+
+                std::string tmp;
+                if (base == 16)
+                    tmp.append("0x");
```

Comment:
What if:
1) base = 0, input = 0x0, msvc = 1
2) base = 0, input = 0x0, msvc = 0
3) base = 16, input = 0x0, msvc = 1
4) base = 16, input = 0x0, msvc = 0
5) base = 16, input = 0, msvc = 1
6) base = 16, input = 0, msvc = 0

does switching msvc on cause change in behavior?

Reply:
No. `0x0` is never parsed as hex, not even with `chars_format::hex`. It will be parsed as `0` with tail `x0`, which we then fail.

_(There are some cases where the compatibility layer may be more permissive than the `from_chars` implementation. There should be no cases where it is less permissive.)_